dataplane/burst.h

```diff
@@ -0,0 +1,124 @@
+#pragma once
+
+#include <rte_ethdev.h>
+#include <rte_mbuf.h>
+
+#include "common.h"
+#include "common/type.h"
+
+namespace dpdk
+{
+
+class BurstSpan
+{
+	rte_mbuf** first_;
+	uint16_t len_ = 0;
+
+public:
+	BurstSpan(rte_mbuf** first, uint16_t length) :
+	        first_{first}, len_{length}
+	{
+	}
+	bool empty() { return len_ == 0; }
+	rte_mbuf** begin()
+	{
+		return first_;
+	}
+	rte_mbuf** end()
+	{
+		return first_ + len_;
+	}
+	uint16_t Free()
+	{
+		if (empty())
+			return 0;
+		rte_pktmbuf_free_bulk(first_, len_);
+		auto freed = len_;
+		len_ = 0;
+		return freed;
+	}
+	uint64_t Bytes()
+	{
+		uint64_t bytes = 0;
+		for (auto& mbuf : *this)
+		{
+			bytes += rte_pktmbuf_pkt_len(mbuf);
+		}
+		return bytes;
+	}
+	uint16_t Packets() { return len_; }
+	BurstSpan& Tx(tPortId port_id, tQueueId queue_id)
+	{
+		if (empty())
+			return *this;
+		auto txSize = rte_eth_tx_burst(port_id, queue_id, first_, len_);
+		len_ -= txSize;
+		first_ += txSize;
+		return *this;
+	}
+	BurstSpan& Tx(tPortId port_id, tQueueId queue_id, uint64_t& opackets, uint64_t& obytes)
+	{
+		if (empty())
+			return *this;
+		uint64_t bytes = Bytes();
+
+		auto txSize = rte_eth_tx_burst(port_id, queue_id, first_, len_);
+
+		opackets += txSize;
+
+		len_ -= txSize;
+		first_ += txSize;
+
+		bytes -= Bytes();
+		obytes += bytes;
+
+		return *this;
+	}
+};
```

Comment:
It really looks like one method can be used in another to avoid code duplication, but I can't think of a solution on top of my head.
The first idea is to use smth like:
```cpp
private:
BurstSpan& Tx(tPortId port_id, tQueueId queue_id, uint64_t* opackets, uint64_t* obytes) {
               if (empty())
			return *this;
		uint64_t bytes = Bytes();
		auto txSize = rte_eth_tx_burst(port_id, queue_id, first_, len_);
		len_ -= txSize;
		first_ += txSize;
		
		if (opackets != nullptr)
		 *opackets += txSize;
		
		if (obytes != nullptr) {
		bytes -= Bytes();
		*obytes += bytes
		}
		return *this
}
public:
BurstSpan& Tx(tPortId port_id, tQueueId queue_id) {
return Tx(port_id, queue_id, nullptr, nullptr);
}
BurstSpan& Tx(tPortId port_id, tQueueId queue_id, uint64_t& opackets, uint64_t& obytes) {
return Tx(port_id, queue_id, &opackets, &obytes);
}
```

Reply:
Suggested solution adds branching on the packet processing pass. This branching is superfluous, considering that decision to count statistics is actually made at compile time.

Reply:
Yes, I know it's not ideal, it was just my first thought to show that we _can_ avoid code duplication.
One _possible_ way is do smth like that:
```cpp
template<typename TPacketPtr = nullptr_t, typename TBytePtr = nullptr_t>
BurstSpan& Tx(tPortId port_id, tQueueId queue_id, TPacketPtr opackets = nullptr, TBytePtr obytes = nullptr) {
    if (empty())
        return *this;
    
    uint64_t bytes = Bytes();
    auto txSize = rte_eth_tx_burst(port_id, queue_id, first_, len_);
    len_ -= txSize;
    first_ += txSize;
    
    if constexpr (!std::is_same_v<TPacketPtr, nullptr_t>) {
        *opackets += txSize;
    }
    
    if constexpr (!std::is_same_v<TBytePtr, nullptr_t>) {
        bytes -= Bytes();
        *obytes += bytes;
    }
    
    return *this;
}

// For cases where opackets and obytes are not used:
BurstSpan& Tx(tPortId port_id, tQueueId queue_id) {
    return Tx<>(port_id, queue_id);
}

// For cases where opackets and obytes are used:
BurstSpan& Tx(tPortId port_id, tQueueId queue_id, uint64_t& opackets, uint64_t& obytes) {
    return Tx(port_id, queue_id, &opackets, &obytes);
}
```

But, I mean, maybe there's a way to just call a smaller function from the bigger one and utilize its return value. I just didn't think about that enough to come up with a good solution.

Reply:
I see that you added TxTracked, I like it.
(I'm still not able to close this conversation by myself though, could you close it for me?)

Reply:
Reworked without output parameters